### PR TITLE
Book crud

### PR DESCRIPTION
**Pull request by hellfishg**
## Summary

Create book CRUD and exceptions

## Trello Card

[trello card] https://trello.com/c/z75Kj68s/18-crear-las-operaciones-crud-de-book

### CODE REVIEW

File: src/main/java/wolox/training/TrainingApplication.java [hunk 1]
```diff
@@ -2,7 +2,11 @@
 
 import org.springframework.boot.SpringApplication;
 import org.springframework.boot.autoconfigure.SpringBootApplication;
+import org.springframework.boot.autoconfigure.domain.EntityScan;
+import org.springframework.data.jpa.repository.config.EnableJpaRepositories;
 
+@EnableJpaRepositories("wolox.training.repositories")
+@EntityScan("wolox.training.models")
```
Comment (jeymarsonwolox) [inline src/main/java/wolox/training/TrainingApplication.java]: this is not necesary

File: src/main/java/wolox/training/controllers/BookController.java [hunk 2]
```diff
@@ -1,16 +1,75 @@
...
+import org.springframework.web.bind.annotation.ResponseStatus;
+import org.springframework.web.bind.annotation.RestController;
+import wolox.training.exceptions.BookNotFoundException;
+import wolox.training.models.Book;
+import wolox.training.repositories.BookRepository;
 
-@Controller
+@RestController
+@RequestMapping("/woloxTraining/books")
 public class BookController {
```
Comment (jeymarsonwolox) [inline src/main/java/wolox/training/controllers/BookController.java:28]: add javadoc

File: src/main/java/wolox/training/exceptions/BookNotFoundException.java [hunk 3]
```diff
@@ -0,0 +1,12 @@
+package wolox.training.exceptions;
+
+import org.springframework.http.HttpStatus;
+import org.springframework.web.bind.annotation.ResponseStatus;
+
+@ResponseStatus(code = HttpStatus.NOT_FOUND , reason = "Book Id Mismatch")
```
Comment (jeymarsonwolox) [inline src/main/java/wolox/training/exceptions/BookNotFoundException.java:10]: add javadoc

File: src/main/java/wolox/training/exceptions/BookAlreadyExistException.java [hunk 4]
```diff
@@ -0,0 +1,12 @@
+package wolox.training.exceptions;
+
+import org.springframework.http.HttpStatus;
+import org.springframework.web.bind.annotation.ResponseStatus;
+
+@ResponseStatus(code = HttpStatus.NOT_ACCEPTABLE, reason = "Book Already Exist")
+public class BookAlreadyExistException extends RuntimeException {
```
Comment (jeymarsonwolox) [inline src/main/java/wolox/training/exceptions/BookAlreadyExistException.java:11]: add javadoc

File: src/main/java/wolox/training/controllers/BookController.java [hunk 5]
```diff
@@ -1,16 +1,75 @@
...
+  @DeleteMapping("/{id}")
+  public void delete(@PathVariable Long id) {
+    Book book = bookRepository.findById(id).orElseThrow(() -> new BookNotFoundException("Book Not Found", new Exception()));
+    bookRepository.delete(book);
+  }
+
+  @PutMapping("/{id}")
+  public Book updateBook(@RequestBody Book book, @PathVariable Long id) {
+    if (book.getId() != id)
+      throw new BookNotFoundException("Book is not found", new Exception());
```
Comment (jeymarsonwolox) [inline src/main/java/wolox/training/controllers/BookController.java]: create `BookIdMismatchException` and use here

File: src/main/java/wolox/training/controllers/BookController.java [hunk 6]
```diff
@@ -1,16 +1,75 @@
...
+    bookRepository.findById(id);
```
Comment (jeymarsonwolox) [inline src/main/java/wolox/training/controllers/BookController.java]: add exception if the book does not exits

**Review by jeymarsonwolox — CHANGES_REQUESTED** [review]

File: src/main/java/wolox/training/controllers/BookController.java [hunk 7]
```diff
@@ -1,16 +1,80 @@
...
+    Book book = bookRepository.findById(id).orElseThrow(() -> new BookNotFoundException("Book Not Found", new Exception()));
+    bookRepository.delete(book);
+  }
+
+  @PutMapping("/{id}")
+  public Book updateBook(@RequestBody Book book, @PathVariable Long id) {
+    if (book.getId() != id) {
+      throw new BookIdMismatchException("That book does not match that id", new Exception());
+    }
+    bookRepository.findById(id).orElseThrow(() -> new BookIdMismatchException("Book is not found", new Exception()));
```
Comment (jeymarsonwolox) [inline src/main/java/wolox/training/controllers/BookController.java]: use `BookNotFoundException` here

**Review by jeymarsonwolox — CHANGES_REQUESTED** [review]

**Review by jeymarsonwolox — APPROVED** [review]

**Review by IgnOliveto — APPROVED** [review]

**Pull request merged**